javatests/arcs/core/data/proto/RecipeProtoDecoderTest.kt

```diff
@@ -0,0 +1,141 @@
+package arcs.core.data.proto
+
+import arcs.core.data.EntityType
+import arcs.core.data.FieldType
+import arcs.core.data.HandleConnectionSpec
+import arcs.core.data.ParticleSpec
+import arcs.core.data.Recipe.Handle
+import arcs.core.data.Recipe.Particle
+import arcs.core.data.Schema
+import arcs.core.data.SchemaName
+import arcs.core.data.SchemaFields
+import arcs.core.data.TypeVariable
+import arcs.core.testutil.assertThrows
+import arcs.core.testutil.fail
+import arcs.core.util.Result
+import arcs.repoutils.runfilesDir
+import com.google.common.truth.Truth.assertThat
+import com.google.protobuf.Message.Builder
+import com.google.protobuf.Message
+import com.google.protobuf.TextFormat
+import java.io.File
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+@RunWith(JUnit4::class)
+class RecipeProtoDecoderTest {
+    // The test environment.
+    var ramdiskStorageKey = "ramdisk://thing"
+    val thingHandleProto = HandleProto
+        .newBuilder()
+        .setName("thing")
+        .setFate(HandleProto.Fate.CREATE)
+        .setStorageKey(ramdiskStorageKey)
+        .build()
+    val thingHandle = Handle(
+        "thing", Handle.Fate.CREATE, ramdiskStorageKey, TypeVariable("thing"), emptyList()
+    )
+    val thingSchema = Schema(
+            names = listOf(SchemaName("Thing")),
+            fields = SchemaFields(singletons=mapOf("name" to FieldType.Text), collections=mapOf()),
+            hash = ""
+    )
+    val thingEntity = EntityType(thingSchema)
+    val readConnectionSpec = HandleConnectionSpec("data", Direction.READS, thingEntity)
+    val readerSpec = ParticleSpec("Reader", mapOf("data" to readConnectionSpec), "ReaderLocation")
+    val writeConnectionSpec = HandleConnectionSpec("data", Direction.WRITES, thingEntity)
+    val writerSpec = ParticleSpec("Writer", mapOf("data" to writeConnectionSpec), "WriterLocation")
+    val context = DecodingContext(
```

Comment:
This probably would be nicer:
```
context = DecodingContext(
  particleSpecs = listOf(readerSpec, writerSpec).associateBy { it.name },
  recipeHandles = listOf(thingHandle).associateBy { it.name }
```
Maybe even the decoding context could accepts collections of specs and handles and do the mapping itself?

Reply:
Good idea about using `associateBy`. Done.

As for changing the decoding context to take a list, having an explicit map keeps the error handling simpler. If we pass the list to the constructor, we need to propagate errors back in case of duplicate names, etc. The `DecodingContext` is used in a very limited way. So, it might be better to keep it as is?